File: bst/levelorder/treedepth/DepthBTIterative.java
package levelorder.treedepth;

import java.util.LinkedList;
import java.util.Queue;


class DepthBTIterative {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int depth = 0;

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            depth++;  // Increase depth for each level

            for (int i = 0; i < levelSize; i++) {
                TreeNode currentNode = queue.poll();
                // Add left and right children to the queue
                if (currentNode.left != null) {
                    queue.offer(currentNode.left);
                }
                if (currentNode.right != null) {
                    queue.offer(currentNode.right);
                }
            }
        }

        return depth;
    }

    public static void main(String[] args) {
        DepthBTIterative solution = new DepthBTIterative();

        // Example 1
        TreeNode root1 = new TreeNode(3);
        root1.left = new TreeNode(9);
        root1.right = new TreeNode(20);
        root1.right.left = new TreeNode(15);
        root1.right.right = new TreeNode(7);
        System.out.println(solution.maxDepth(root1));  // Output: 3

        // Example 2
        TreeNode root2 = new TreeNode(1);
        root2.right = new TreeNode(2);
        System.out.println(solution.maxDepth(root2));  // Output: 2
    }
}

File: bst/levelorder/treedepth/DepthBTIterative.py
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class DepthBTIterative:
    def maxDepth(self, root: TreeNode) -> int:
        if not root:
            return 0

        queue = deque([root])
        depth = 0

        # Perform BFS
        while queue:
            depth += 1
            level_size = len(queue)  # Number of nodes at the current level
            for _ in range(level_size):
                current = queue.popleft()
                # Add the child nodes for the next level
                if current.left:
                    queue.append(current.left)
                if current.right:
                    queue.append(current.right)

        return depth

# Example usage:
if __name__ == "__main__":
    solution = DepthBTIterative()

    # Example 1
    root1 = TreeNode(3)
    root1.left = TreeNode(9)
    root1.right = TreeNode(20)
    root1.right.left = TreeNode(15)
    root1.right.right = TreeNode(7)
    print(solution.maxDepth(root1))  # Output: 3

    # Example 2
    root2 = TreeNode(1)
    root2.right = TreeNode(2)
    print(solution.maxDepth(root2))  # Output: 2
